beacon-chain/state/stategen/replayer.go

```diff
@@ -0,0 +1,347 @@
+			// err will be handled after the bookend log
+			s, err = ReplayProcessSlots(ctx, s, target)
+		}
+
+		duration := time.Since(start)
+		log.WithFields(logrus.Fields{
+			"duration": duration,
+		}).Debug("time spent in process_slots")
+		if err != nil {
+			return nil, errors.Wrap(err, fmt.Sprintf("ReplayToSlot failed to seek to slot %d after applying blocks", target))
+		}
+	}
+	return s, nil
+}
+
+// ReplayerBuilder creates a Replayer that can be used to obtain a state at a specified slot or root
+// (only ForSlot implemented so far).
+// See documentation on Replayer for more on how to use this to obtain pre/post-block states
+type ReplayerBuilder interface {
+	// ForSlot creates a builder that will create a state that includes blocks up to and including the requested slot
+	// The resulting Replayer will always yield a state with .Slot=target; if there are skipped blocks
+	// between the highest canonical block in the db and the target, the replayer will fast-forward past the intervening
+	// slots via process_slots.
+	ForSlot(target types.Slot) Replayer
+}
+
+// NewCanonicalBuilder handles initializing the default concrete ReplayerBuilder implementation.
+func NewCanonicalBuilder(h HistoryAccessor, c CanonicalChecker, cs CurrentSlotter) *CanonicalBuilder {
+	return &CanonicalBuilder{
+		Chainer: &CanonicalChainer{
+			h:  h,
+			c:  c,
+			cs: cs,
+		},
+	}
+}
+
+type retrievalMethod int
+
+const (
+	forSlot retrievalMethod = iota
+	byRoot
+)
+
+// CanonicalBuilder builds a Replayer that uses a combination of database queries and a
+// CanonicalChecker (which should usually be a fork choice store implementing an IsCanonical method)
+// to determine the canonical chain and apply it to generate the desired state.
+type CanonicalBuilder struct {
+	Chainer Chainer
+}
+
+var _ ReplayerBuilder = &CanonicalBuilder{}
+
+func (r *CanonicalBuilder) ForSlot(target types.Slot) Replayer {
+	return &stateReplayer{chainer: r.Chainer, method: forSlot, blockSlot: target}
+}
+
+// Chainer is responsible for supplying the chain components necessary to rebuild a state,
+// namely a starting BeaconState and all available blocks from the starting state up to and including the target slot
+type Chainer interface {
+	ChainForSlot(ctx context.Context, target types.Slot) (state.BeaconState, []block.SignedBeaconBlock, error)
+}
+
+type CanonicalChainer struct {
+	h  HistoryAccessor
+	c  CanonicalChecker
+	cs CurrentSlotter
+}
+
+var _ Chainer = &CanonicalChainer{}
+
+// ChainForSlot creates a value that satisfies the Replayer interface via db queries
+// and the stategen transition helper methods. This implementation uses the following algorithm:
+// - find the highest canonical block >= the target slot
+// - starting with this block, recursively search backwards for a stored state, and accumulate intervening blocks
+func (r *CanonicalChainer) ChainForSlot(ctx context.Context, target types.Slot) (state.BeaconState, []block.SignedBeaconBlock, error) {
```

Comment:
Do we really need to export it?

```suggestion
func (r *CanonicalChainer) chainForSlot(ctx context.Context, target types.Slot) (state.BeaconState, []block.SignedBeaconBlock, error) {
```

Reply:
agree, making this private now that it isn't needed for tests outside this package